relay_client/src/client/fetch.rs

```diff
@@ -0,0 +1,104 @@
+use {
+    crate::{create_request, Client, Error, ResponseFuture},
+    futures_util::{FutureExt, Stream},
+    relay_rpc::{
+        domain::Topic,
+        rpc::{BatchFetchMessages, SubscriptionData},
+    },
+    std::{
+        pin::Pin,
+        task::{Context, Poll},
+    },
+};
+
+/// Stream that uses the `irn_batchFetch` RPC method to retrieve messages from
+/// the Relay.
+pub struct FetchMessageStream {
+    client: Client,
+    request: BatchFetchMessages,
+    batch: Option<std::vec::IntoIter<SubscriptionData>>,
+    batch_fut: Option<ResponseFuture<BatchFetchMessages>>,
+    has_more: bool,
+}
+
+impl FetchMessageStream {
+    pub(super) fn new(client: Client, topics: impl Into<Vec<Topic>>) -> Self {
+        let request = BatchFetchMessages {
+            topics: topics.into(),
+        };
+
+        Self {
+            client,
+            request,
+            batch: None,
+            batch_fut: None,
+            has_more: true,
+        }
+    }
+
+    /// Clears all internal state so that on the next stream poll it returns
+    /// `None` and finishes data streaming.
+    #[inline]
+    fn clear(&mut self) {
+        self.batch = None;
+        self.batch_fut = None;
+        self.has_more = false;
+    }
+}
+
+impl Stream for FetchMessageStream {
+    type Item = Result<SubscriptionData, Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
```

Comment:
I'd name context variable `ctx` (otherwise it looks a bit like `tx`, `rx` thing).

Reply:
Yeah, but it's auto-generated from method signature. For some reason they named it `cx`...